core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java

```diff
@@ -955,10 +955,38 @@ static RelDataType deriveTypeSplit(SqlOperatorBinding operatorBinding,
           .withKind(SqlKind.CONCAT_WS_MSSQL);
 
   private static RelDataType arrayReturnType(SqlOperatorBinding opBinding) {
-    RelDataType type =
-        opBinding.getOperandCount() > 0
-            ? ReturnTypes.LEAST_RESTRICTIVE.inferReturnType(opBinding)
-            : opBinding.getTypeFactory().createUnknownType();
+    // only numeric & character types check
+    List<RelDataType> numericTypes = new ArrayList<>();
+    List<RelDataType> characterTypes = new ArrayList<>();
+    List<RelDataType> nullTypes = new ArrayList<>();
+    for (int i = 0; i < opBinding.getOperandCount(); i++) {
+      SqlTypeFamily family = opBinding.getOperandType(i).getSqlTypeName().getFamily();
+      if (family == SqlTypeFamily.NUMERIC) {
+        numericTypes.add(opBinding.getOperandType(i));
+      } else if (family == SqlTypeFamily.CHARACTER) {
+        characterTypes.add(opBinding.getOperandType(i));
+      } else if (family == SqlTypeFamily.NULL) {
+        nullTypes.add(opBinding.getOperandType(i));
+      } else {
+        break;
+      }
+    }
+
+    boolean onlyNumericCharacterTypes = opBinding.getOperandCount() > 0
```

Comment:
I could be misunderstanding, but if you only care about the size of the lists and not the elements themselves, you could just use boolean flags to keep track of which types have appeared. I think it will simplify the expression below.

Reply:
```     final List<RelDataType> types = opBinding.collectOperandTypes();
    boolean hasNumeric = false;
    boolean hasCharacter = false;
    types.forEach(type -> {
			// null check here maybe
			switch (type.getSqlTypeName().getFamily())  {
			 case x:
			 		hasNumeric = true;
			 		break;
			 case y:
			 		hasCharacter = true;
			 		break;
			 default:
		 }
		 if hasNumeric || hasCharacter || c {....} ``` you could do something like this maybe? (Excuse the pseudocode it's just an example)

Reply:
good point. I have fixed this. PTAL.